### PR TITLE
Add sql directory parser and dag generator

### PR DESCRIPTION
**Pull request by feluelle**
# Description

## What is the current behavior?

Currently, we do not have a way of parsing a directory of sql files and generating a DAG out of it.

closes: #794
related: #897

## What is the new behavior?

This PR adds a _SQL Directory Parser_ as well as a _DAG Generator_ with examples.

The current implementation adds minimal configurability. More can be handled in a separate PR.
The following assumptions have been made:

1. The sql files **directory** can be passed via cli arg
1. The **start_date** is `datetime.now()`
1. The **conn_id**, **schema** and **database** can be set via sql file header
1. The **output table name** equals the sql file name

## Does this introduce a breaking change?

No.

### Checklist
- [x] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

**Comment by feluelle** [comment]
This is a **very early draft** of how it could look like using multiple `aql.transform_file` operators.

cc @dimberman

**Comment by feluelle** [comment]
Failed checks are unrelated. I'll add unit tests once we agree to a specific strategy.

I tested it by installing the following deps:
* _Jinja2_ (very useful for templating and generating the dag)
* _fs_ (optional but useful for working with file systems)
* _astro-sdk_
* _networkx_ (for detecting the right order in which to call the transform steps)

You can then run the example via `AIRFLOW__CORE__ENABLE_XCOM_PICKLING=True python sql-cli/src/sql_cli/parser.py`.

**Comment by feluelle** [comment]
@dimberman @tatiana PTAL

Note that we can add meta data to the sql files and use `frontmatter` as done by Daniel to read it. Or we use yaml file(s) for configuration and additional meta data. But please let me know first if this approach, using _jinja_ and _networkx_, could be a potential candidate OR we use what Daniel [has built](https://github.com/astronomer/astro-sdk/blob/0.8.4/src/astro/sql/parsers/sql_directory_parser.py) already (`aql.render`).

I like the jinja approach more as it generates a dag file from a template once (e.g. on cli call) which should in the end be faster for the airflow dag parser to process - instead of the `aql.render` which generates the dag/tasks every time the DAG file calling it gets parsed. Moreover the jinja approach allows someone to debug the DAG code if necessary whereas in the `aql.render` approach everything is dynamically being generated.

**Comment by kaxil** [comment]
Markdown link check should be fixed by https://github.com/astronomer/astro-sdk/pull/842

File: sql-cli/src/sql_cli/sql_directory_parser.py
Comment (tatiana) [inline sql-cli/src/sql_cli/sql_directory_parser.py:45]: If the `SQL` files belong to subdirectories of the given path, are we happy to have all the tasks flattened in the DAG, disregarding the (sub-)directory structure?
Would it make sense to use task groups to represent those inner groups?

Example of SQL directory
```
sql_dir/
├── sub_dir_1
│       ├── a.sql
│       └── b.sql
├── sub_dir_2
│       └── c.sql
│       └── d.sql
├── e.sql
└──  f.sql
```

File: sql-cli/src/sql_cli/dag_generator.py
Comment (tatiana) [inline sql-cli/src/sql_cli/dag_generator.py:29]: Would it be possible for us to detail where the circular dependency is, to helping the user to troubleshoot the issue?

File: sql-cli/src/sql_cli/dag_generator.py
Comment (tatiana) [inline sql-cli/src/sql_cli/dag_generator.py:3]: Networkx is a very mature library, and I think you used it well.
Do you know how Airflow identifies cyclic dependencies at the moment? I'd be curious to see how both approaches relate.

**Comment by tatiana** [comment]
@feluelle I like how simple and how explicit this approach is. Let's wait for other's feedback :crossed_fingers:

File: sql-cli/src/sql_cli/sql_directory_parser.py
Comment (feluelle) [reply sql-cli/src/sql_cli/sql_directory_parser.py:45]: I had the same idea. We can indeed reconstruct the directory structure in Airflow using TaskGroups. But I would say this can be a separate feature.

File: sql-cli/src/sql_cli/dag_generator.py
Comment (feluelle) [reply sql-cli/src/sql_cli/dag_generator.py:3]: I don't know, but I can try to find out. @kaxil do you know?

File: sql-cli/src/sql_cli/dag_generator.py
Comment (feluelle) [reply sql-cli/src/sql_cli/dag_generator.py]: Resolved in cb8f6de

File: sql-cli/src/sql_cli/dag_generator.py
Comment (dimberman) [inline sql-cli/src/sql_cli/dag_generator.py:3]: Interesting touch to using a network analysis library for creating the SQL graphs. A few questions:

1. How big is this library? 
2. Are we doing complex enough DAG creation to require this library?

I don't have strong thoughts either way, just want to make sure we're asking the question. Certainly seems to make the code cleaner :).

File: sql-cli/src/sql_cli/dag_generator.py
Comment (dimberman) [inline sql-cli/src/sql_cli/dag_generator.py:25]: This line seems... wrong? Like we need to make sure that the variable referenced is actually a table before we add it to the graph.

File: sql-cli/src/sql_cli/sql_directory_parser.py
Comment (dimberman) [reply sql-cli/src/sql_cli/sql_directory_parser.py:43]: I think reconstructing using taskgroups makes sense, it works nicely in gusty

**Review by dimberman — CHANGES_REQUESTED** [review]
@feluelle given the pros and cons we described, are we only going with the "generate transform_file function" option? Or would the idea be that the user can decide whether to generate or simply render at runtime?

I think given the speeds we've seen I don't think that "it renders faster" is a huge concern here. We were able to render 1000 tasks in pretty much the same amount of time. 

The debugging argument also seems a bit iffy to me. Ultimately the error either way is going to be "could not parse at ..../foo.sql.py", I don't think it really matters which line in the airflow DAG starts the task. Either way the user is going to need to debug the SQL file itself.

Also with the `aql.transform_file` approach the DAG would be essentially unreadable when you get too 100 or 1000 tasks. It would generate a monstrously large DAG that would be really painful to edit.

File: sql-cli/src/sql_cli/dag_generator.py
Comment (feluelle) [reply sql-cli/src/sql_cli/dag_generator.py]: In 2.4.0, [this](https://github.com/apache/airflow/blob/2.4.0/airflow/utils/dag_cycle_tester.py#L49-L83) is how it detects cycles. So it does not use any external package for that.

**Comment by feluelle** [comment]
> @feluelle given the pros and cons we described, are we only going with the "generate transform_file function" option? Or would the idea be that the user can decide whether to generate or simply render at runtime?

I think it makes sense to offer both options:
* `aql.render` in python-sdk as it is a python function which can directly be added to an existing DAG
  * probably being favored by DE
* and the jinja approach in sql-cli as it does not require the user to know airflow or python at all to generate a DAG
  * probably being favored by DA

> I think given the speeds we've seen I don't think that "it renders faster" is a huge concern here. We were able to render 1000 tasks in pretty much the same amount of time.

This is correct. I have edited my comment in GitHub to link to the notion doc.

> The debugging argument also seems a bit iffy to me. Ultimately the error either way is going to be "could not parse at ..../foo.sql.py", I don't think it really matters which line in the airflow DAG starts the task. Either way the user is going to need to debug the SQL file itself.

Same here. Please let us have this discussion in [this comment](https://www.notion.so/Parsing-a-SQL-directory-as-a-DAG-876112c7646349fdb09c654711d38328?d=9e59a5273e144f39a435239d981d3375#b43f2720812f444d80e6e0a52b5e2c91) of the docs.

> Also with the `aql.transform_file` approach the DAG would be essentially unreadable when you get too 100 or 1000 tasks. It would generate a monstrously large DAG that would be really painful to edit.

The point is that you should not edit it as it gets overwritten every time we regenerate the DAG file. This is indeed one of the "Cons" of this approach, see this [block](https://www.notion.so/astronomerio/Parsing-a-SQL-directory-as-a-DAG-876112c7646349fdb09c654711d38328#01b39ca6254c4b9e8e01145e5908c7a2) in the docs. With the new dataset support in Airflow and astro-sdk, we can also easily split it into multiple DAGs 👍 (we should probably do that anyway, even if not meant to be edited.)

File: sql-cli/src/sql_cli/dag_generator.py
Comment (feluelle) [reply sql-cli/src/sql_cli/dag_generator.py:3]: Yes, totally fair to ask this question.

I personally dont trust my graph algorithm skills enough to reimplement this from scratch. 😅 I am afraid, I am missing edge cases etc. That is why I decided to use _networkx_. It has builtin functions for cycle detection and finding the right order to execute statements in using dfs.

File: sql-cli/src/sql_cli/dag_generator.py
Comment (feluelle) [reply sql-cli/src/sql_cli/dag_generator.py:25]: Not sure what you mean. Every `parameter` is an input table and `sql_file.get_variable_name()` [equals](https://github.com/astronomer/astro-sdk/blob/cb8f6de7c6c64130f8c04a11798bcf6d23a5e1d8/sql-cli/src/sql_cli/template.jinja2#L17) the output table name in the **current** implementation.

File: sql-cli/src/sql_cli/sql_directory_parser.py
Comment (feluelle) [reply sql-cli/src/sql_cli/sql_directory_parser.py]: Added in 07da4f9

**Comment by feluelle** [comment]
@dimberman @tatiana I changed the code to use `transform_file` but with raw sql code instead of file path, because of two reasons:

1. If you develop locally the file path will might be different to where you deploy to e.g. I generated the DAG locally but want to run it in astro-cli (in docker). I would have to copy the files or create a volume to be able to reference the files
2. Because you can see the raw sql code being deployed - no surprises! If I remember correctly, I think this was also a request from Mike. I also don't think we can see the sql code in the "Rendered" view of the task instance if we pass a file path. With the current implementation you can see the rendered sql code after #897 has been merged.

**Another note:**
I can also create a `transform_raw` function or similar (unfortunately `transform` we already have, so I would have to move it somewhere else if we want to use the same name `transform`) to make it clearer that we transform raw sql - not sql from a file.

File: sql-cli/examples/_dags/advanced.py
Comment (dimberman) [inline sql-cli/examples/_dags/advanced.py:36]: @feluelle this isn't a file path, it's a SQL function, is this an error?

File: sql-cli/examples/_dags/advanced.py
Comment (dimberman) [inline sql-cli/examples/_dags/advanced.py:16]: @feluelle if our ultimate goal with transform_file over render is readability, I think we need to generate better names. I don't think a user would be able to parse what any of this means.

Comment (dimberman) [reply sql-cli/examples/_dags/advanced.py:16]: I think that the name should be generated based on the file name, not the taskgroup.

Also why are these tasks under two task groups?

File: sql-cli/examples/_dags/advanced.py
Comment (dimberman) [inline sql-cli/examples/_dags/advanced.py:62]: why are we outputing to a table named "a"? Either the tables should have names that mean something or we should use temporary tables.

File: sql-cli/examples/_dags/simple.py
Comment (dimberman) [inline sql-cli/examples/_dags/simple.py:26]: again, if we're using "transform_file", why are we passing in SQL statements instead of file paths?

File: sql-cli/examples/_dags/simple.py
Comment (dimberman) [inline sql-cli/examples/_dags/simple.py:11]: Assuming this is based on the file name, we should have better file names so that our variables have more interesting names. Single letter variables in our example DAGs won't give an impression of production readiness.

File: sql-cli/examples/advanced/sub_dir_1/a.sql
Comment (dimberman) [inline sql-cli/examples/advanced/sub_dir_1/a.sql:1]: This is code that will be seen by our users if they visit our GitHub. We should come up with better file names than "a.sql", "b.sql", etc.

File: sql-cli/src/sql_cli/dag_generator.py
Comment (dimberman) [inline sql-cli/src/sql_cli/dag_generator.py:7]: Can this same class be used as the backend for an eventual render function as well?

Maybe we can factor out certain elements of this and put it into the python-sdk module for that function?

File: sql-cli/src/sql_cli/dag_generator.py
Comment (dimberman) [inline sql-cli/src/sql_cli/dag_generator.py:3]: Huge +1 for having an external library handle graph creation. We should scale test this with some different configurations to make sure there's no efficiency issues but otherwise I'm always a fan of offloading work.

File: sql-cli/src/sql_cli/dag_generator.py
Comment (dimberman) [inline sql-cli/src/sql_cli/dag_generator.py:20]: We should only be creating edges for parameters that are of type "table"

File: sql-cli/src/sql_cli/sql_directory_parser.py
Comment (dimberman) [inline sql-cli/src/sql_cli/sql_directory_parser.py:10]:
```suggestion
    A SqlFile is equivalent to a transform step in the Astro SDK.
```

File: sql-cli/src/sql_cli/sql_directory_parser.py
Comment (dimberman) [inline sql-cli/src/sql_cli/sql_directory_parser.py:129]: Should we create a taskgroup based on the directory name for nested directories?

**Review by dimberman — CHANGES_REQUESTED** [review]

File: sql-cli/examples/_dags/advanced.py
Comment (feluelle) [reply sql-cli/examples/_dags/advanced.py:36]: Please read my https://github.com/astronomer/astro-sdk/pull/836#issuecomment-1256174948 above (with the note).

File: sql-cli/examples/_dags/advanced.py
Comment (feluelle) [reply sql-cli/examples/_dags/advanced.py:16]: > @feluelle if our ultimate goal with transform_file over render is readability, I think we need to generate better names. I don't think a user would be able to parse what any of this means.

This automatically generated code from the file names and the directory names. We can change it easily. I agree with you that we should not use this in official examples in our repository. Do you have some examples in mind? :)

> I think that the name should be generated based on the file name, not the taskgroup.

Then it won't be possible to use the same filename twice. That is what I want to show with this example. It contains two `a.sql` in different folders/task groups.

> Also why are these tasks under two task groups?

Because the file is in a **nested** directory, it will be in nested Task Groups. I think this is pretty useful when you have lots of tables to be able to create nested structures.

File: sql-cli/examples/_dags/advanced.py
Comment (feluelle) [reply sql-cli/examples/_dags/advanced.py:62]: Because currently, the table name equals the file name as in dbt. I am completely open to use sql headers or yaml for other and more configuration options.

File: sql-cli/examples/_dags/simple.py
Comment (feluelle) [reply sql-cli/examples/_dags/simple.py:26]: See https://github.com/astronomer/astro-sdk/pull/836#discussion_r978882933

File: sql-cli/examples/_dags/simple.py
Comment (feluelle) [reply sql-cli/examples/_dags/simple.py:11]: I agree with you. This was only for testing purpose. I need to come up with some more realistic names. If you have any suggestion, please let me know :)

File: sql-cli/src/sql_cli/dag_generator.py
Comment (feluelle) [reply sql-cli/src/sql_cli/dag_generator.py:7]: Interesting idea. 🤔

File: sql-cli/src/sql_cli/dag_generator.py
Comment (feluelle) [reply sql-cli/src/sql_cli/dag_generator.py:20]: The "if" one line below checks that by doing this `if parameter in self.nodes  # only add edges for existing nodes`. `self.nodes` only contains file names / input tables.

File: sql-cli/src/sql_cli/sql_directory_parser.py
Comment (feluelle) [reply sql-cli/src/sql_cli/sql_directory_parser.py:129]: Is this a feature request? Or a rhetorical question? 😅 The current implementation via `rglob("*.sql")` does a recursive traversal. So we find **all** directories.

I think yes. I think it makes sense to reflect the directory structure in the database as schemas.

**Comment by feluelle** [comment]
@dimberman @tatiana please see [this slack thread](https://astronomer.slack.com/archives/C042ZUJ7QG2/p1664181739485199) about why I had to remove the TaskGroup feature.

File: sql-cli/examples/_dags/advanced.py
Comment (feluelle) [reply sql-cli/examples/_dags/advanced.py]: fixed in 3372688

File: sql-cli/examples/_dags/advanced.py
Comment (feluelle) [reply sql-cli/examples/_dags/advanced.py]: fixed in c6902aa

File: sql-cli/examples/_dags/simple.py
Comment (feluelle) [reply sql-cli/examples/_dags/simple.py]: fixed in c6902aa

File: sql-cli/examples/advanced/sub_dir_1/a.sql
Comment (feluelle) [reply sql-cli/examples/advanced/sub_dir_1/a.sql]: fixed in c6902aa

File: .pre-commit-config.yaml
Comment (tatiana) [inline .pre-commit-config.yaml:10]: Is there anything about the DAGs we are generating which MyPy doesn't like?

File: sql-cli/examples/_dags/simple.py
Comment (tatiana) [inline sql-cli/examples/_dags/simple.py:1]: It would be great to have tests running validating the DAG Generator works as expected. I wonder if the content of the `examples` directory should be within tests. We could have an integration test that inputs SQL files and outputs the DAG file and checks if the DAG file is what we want it to be.

File: sql-cli/examples/advanced/mart/union_top_and_last.sql
Comment (tatiana) [inline sql-cli/examples/advanced/mart/union_top_and_last.sql:1]: I see these more as test fixtures than examples, from a `sql-cli` project perspective, WDYT @feluelle ?

File: .pre-commit-config.yaml
Comment (feluelle) [reply .pre-commit-config.yaml:10]: No, but I think it makes sense to show the raw generated example dags to users instead of formatting them.

But yeah, we can run lint, I guess.

File: sql-cli/examples/advanced/mart/union_top_and_last.sql
Comment (feluelle) [reply sql-cli/examples/advanced/mart/union_top_and_last.sql]: But these are the same examples we have for `python-sdk`. I think it makes sense to have some examples in an `examples` directory - other than tests. For tests we can also use shorter variables such as `a`, `b`, ... 😄

File: sql-cli/src/sql_cli/dag_generator.py
Comment (feluelle) [reply sql-cli/src/sql_cli/dag_generator.py:7]: I am constantly refactoring the code. Let's have a look again when it is a bit more mature.

**Comment by feluelle** [comment]
@dimberman @tatiana I have updated my code and also added lots of tests. PTAL when you have time. Thank you.

File: sql-cli/src/sql_cli/dag_generator.py
Comment (kaxil) [inline sql-cli/src/sql_cli/dag_generator.py:8]: Should we have a separate file for custom exceptions?

File: sql-cli/src/sql_cli/main.py
Comment (kaxil) [inline sql-cli/src/sql_cli/main.py:15]: This is temporary until we switch to click?

File: sql-cli/src/sql_cli/main.py
Comment (kaxil) [inline sql-cli/src/sql_cli/main.py:18]: Docstrings for all the function please

File: sql-cli/src/sql_cli/sql_directory_parser.py
Comment (kaxil) [inline sql-cli/src/sql_cli/sql_directory_parser.py:44]: curious on using the "name" for ">" -- what's the rationale behind it? Depending on that, we might want to use file-size

File: sql-cli/src/sql_cli/sql_directory_parser.py
Comment (kaxil) [inline sql-cli/src/sql_cli/sql_directory_parser.py:30]: why `__o` instead of just `other`?

Comment (kaxil) [reply sql-cli/src/sql_cli/sql_directory_parser.py:30]: similar to what you already have in `__gt__`

**Review by kaxil — APPROVED** [review]
I have added some minor comments

File: sql-cli/src/sql_cli/dag_generator.py
Comment (feluelle) [reply sql-cli/src/sql_cli/dag_generator.py:8]: Makes sense. We could have more exceptions in the future.

File: sql-cli/src/sql_cli/main.py
Comment (feluelle) [reply sql-cli/src/sql_cli/main.py:15]: This is correct.

File: sql-cli/src/sql_cli/main.py
Comment (feluelle) [reply sql-cli/src/sql_cli/main.py:18]: Sure. :)

File: sql-cli/src/sql_cli/sql_directory_parser.py
Comment (feluelle) [reply sql-cli/src/sql_cli/sql_directory_parser.py:44]: I will document it better even-though I pass the comparison to downstream _pathlib.Path_ comparison. It is about sorting based on name - not size.

File: sql-cli/src/sql_cli/sql_directory_parser.py
Comment (feluelle) [reply sql-cli/src/sql_cli/sql_directory_parser.py:30]: This was auto-generated for me when I had typed `def __eq` in my vscode 😅. I have updated it.

**Comment by feluelle** [comment]
The failing checks are unrelated. I will try to address them separately.

File: .pre-commit-config.yaml
Comment (tatiana) [inline .pre-commit-config.yaml:25]: Probably worth adding a comment why we're excluding this.

File: sql-cli/pyproject.toml
Comment (tatiana) [inline sql-cli/pyproject.toml:20]: This can also go away, since it is a dependency of `astro-sdk-python`

File: sql-cli/src/sql_cli/dag_generator.py
Comment (tatiana) [inline sql-cli/src/sql_cli/dag_generator.py:39]: :raised_hands: amazing! thanks for adding this

**Review by dimberman — APPROVED** [review]
This looks fantastic @feluelle. well-documented, clean, love it.

File: sql-cli/src/sql_cli/macros/tasks.py.jinja2
Comment (tatiana) [inline sql-cli/src/sql_cli/macros/tasks.py.jinja2:20]: I like that we're being explicit at first, even though it may be verbose, it will help us - and the end-users - troubleshooting any issues.

File: sql-cli/src/sql_cli/sql_directory_parser.py
Comment (tatiana) [inline sql-cli/src/sql_cli/sql_directory_parser.py:130]: This is helpful to avoid cycles with symbolic links - it may be worth adding a test for this (it can be done ina  separate PR)

File: sql-cli/src/sql_cli/utils.py
Comment (tatiana) [inline sql-cli/src/sql_cli/utils.py:1]: In future, we may be able to rename this module to something jinja related - so we avoid generic utils blowing up. I'm happy for us to do this in a separate PR as well :)

File: sql-cli/tests/_dags/sql_files.py
Comment (tatiana) [inline sql-cli/tests/_dags/sql_files.py:1]: This could go to `.gitignore` (as well as `tests/target`)

**Review by tatiana — APPROVED** [review]
@feluelle minor comments, looking forward to seeing the feedback from our end-users on 0.1! :tada:

File: sql-cli/src/sql_cli/sql_directory_parser.py
Comment (feluelle) [reply sql-cli/src/sql_cli/sql_directory_parser.py]: This did not check for symlinks, but [b7aeae8](https://github.com/astronomer/astro-sdk/pull/836/commits/b7aeae856e5d8e7e76e08976319e4ccae7bdedf1) does.

**Pull request merged**
